example/tpm_ibmtss2/README.md

```diff
@@ -0,0 +1,92 @@
+Test script for TPM 2
+
+Usage
+-----
+
+The TPM 2 test script loads the RSA key into the TPM's NULL hierarchy. Please
+make sure that the NULL hierarchy has no password set and no other
+application is using the TPM 2 and sufficient space is available in the TPM for
+loading the RSA key. Since access to TPM devices is typically restricted to the
+root user, the test script (time_decrypt) will need to be run as root.
+
+The TPM 2 test script can for example be run against swtpm using the vtpm
+proxy device:
+
+ > modprobe tpm_vtpm_proxy
+ > swtpm chardev --vtpm-proxy --tpmstate dir=./ --tpm2
+ New TPM device: /dev/tpm1 (major/minor = 253/1)
+
+For the test script to use the proper device, which would be /dev/tpmrm1,
+set the following environment variable:
+
+ TPM_DEVICE=/dev/tpmrm1
+
+
+When using a hardware TPM 2 the following should be used:
+
+ TPM_DEVICE=/dev/tpmrm0
+
+The following packages are needed:
+
+ - openssl-devel / libssl-dev
+ - tss2-devel / libtss2-dev
+ - swtpm-tools
+
+Run `step0.sh`, `step1.sh` as normal. Instead of running `step2.sh` run
+the `step2-marvin.sh` script.
+
+Compile this reproducer:
+```
+gcc -o time_decrypt time_decrypt.c -lcrypto -libmtss
+```
+
+Execute it against one of the `pms_values.bin` files, for either 1024
+(-n 128) or 2048 bit (-n 256):
+```
+TPM_DEVICE=/dev/tpmrm1 ./time_decrypt -i rsa2048_repeat/pms_values.bin \
+-o rsa2048_repeat/raw_times.bin -k rsa2048/pkcs8.pem -n 256
+```
+
+Note:
+If swtpm is used with an 'older' version of OpenSSL, then the TPM 2 will
+return error messages when a decryption failure occurs (e.g., bad padding)
+and the output will show the following:
+
+```
+finished  total: 22000 decryption failures: 16000  unexpected msg len: 0
+```
+
+Further, newer versions of OpenSSL will not indicate the decryption failure but
+return a message anyway, though its length will be unexpected or may have bad
+content (with expected length):
```

Comment:
that's not what should happen... if the data was correctly created with the `step2-marvin.sh`, then all ciphertexts will decrypt to exact same length

Reply:
1024 bit keys decrypt everything but something is wrong with 2048...

Reply:
Putting this PR in draft mode until I have figured out what's wrong with the TPM (I think) when 2048 bit RSA keys are used.

Reply:
I am having a bit of a mystery with 2048bit keys. I isolated the first encrypted messages created with step2-marvin.sh. Trying to decrypt it wtih openssl 1.1.1l assuming pkcs padding leads to this:

```
# openssl rsautl -inkey rsa2048/key.pem -decrypt -in enc.bin -pkcs
RSA operation error
[...]
```

So the input is bad and indeed raw decryption shows that there's no pkcs#1 v1.5 type of encryption padding.

Trying this now on openssl 3.0.9 leads to this:
```
# openssl rsautl -inkey rsa2048/key.pem -decrypt -in enc.bin -pkcs | od -tx1
The command rsautl was deprecated in version 3.0. Use 'pkeyutl' instead.
0000000 c6 4e b7 9a 31 c9 ba 2f ea 90 88 e3 94 74 75 58
0000020 70 96 3c 1d a4 61 8c 8b 2d b3 1d ba 61 2c d4 d4
0000040 91 2b 10 99 2d 48 3f 9f 93 f3 c3 0a 16 62 3c eb
0000060 ec c7 64 62 a7 90 03 2e 6b 39 18 79 74 24 9a fa
0000100 fa b3 c3 67 a8 47 5b e4 c7 dc 3e 3c 64 db 57 b0
0000120 64 59 ac a2 3c ce e3 eb 8a 9f 08 d6 cc 15 66 88
0000140 bc fe 5b 14 ae 4f 59 ba 11 83 d6 10 81 72 1d 49
0000160 4d f3 85 a7 ba fd e4 31 50 4c 92 29 1d 8b 62 cc
0000200 a0 62 2a 1c a1 35 6f 8d 62 f7 dc 4a 16 9d 55 fb
0000220 f8 87 0a 45 40 7c f8 85 8b dd a9 d8 10 52 60 d1
0000240 40 cc c3 d7 5d 36 30 27 0a
0000251
```

That's 169 bytes returned.

What I don't understand is that the openssl test case seems to succeed on the decryption but then also results in plaintext_len = 48 -- for every input! How can this be? When I run this inside the TPM 2 code on the same machine with same openssl version I get 149 bytes back.

https://github.com/tomato42/marvin-toolkit/blob/d85dfe635c7b28ae582cb0239ddbd1c2b25f6d28/example/openssl/time_decrypt.c#L231-L245


key:
```
-----BEGIN PRIVATE KEY-----
MIIEvgIBADANBgkqhkiG9w0BAQEFAASCBKgwggSkAgEAAoIBAQCnszGhwO5/fTFy
HvWuY+zrc0SOZWUCuEmEpyTwOi4+Xlhfn+Wrq5Lrso8h0yL9E/s8YoB83DkeqvIy
rFKy3nGjCKLpEz902krglJkh4HgRSqFqNOyvscwcgjFObnabOonTzLLFFZDsrFUR
hvBZDuxXPOf1oQe2EFt2MKgm/VHnOxNO7IjbSwMDoP+OKAimwmvsQyFsCIn5edgg
t4MnotZPyatx5iwrMevPVclx0aa6zSDGIdfXRxbFCWM6gUreFNd0zMuodCXyNAbe
+KRDldBBHvbaceVYFes+hiRbk39NkZUmSn9McanV6+cTozgNbgzDi69SusQyT/ZA
Ixp25xLZAgMBAAECggEAC7Rc8tFWjBEit/h7dQUeKnZIewhOZyTR48dvM3GaG+oF
kMIGG4fdeKQlK4sTGncQLlAhL+XGelKb1ifTG08dfw+poCsi3w7j0j4o3HWQZ8ON
1c31FZOrH3PCSXs5/b5Vk5HqSclmLLCf/NnOUWbyXanfdjQ7p6xHV+3QNK1yT5lB
3p925V9+WEDhK1Xq1vbFr2RJa4PTXhg9epdjx0uTOdLJy14zUqsc1brxXvr2brAr
Jk5wipu1aoO+rEF6B2diqxUKBAz7pzpZJIFxleOZxz8dtYtf+k8IU5+hS9aIx4OD
36DEhA4zpmPOI16ORv4ohEKyx0dhbL0MI1gf0PL+EQKBgQDX1rTN6D6HchG+mGW7
0aQjaVjF0YZB+vUFmhw/v/mUvBq88Uwr4qYQ9z5WkaWPgiUX3ejDIR/V0CjONtF5
Q6AC77xP2niXA3xYhTJl6/zjeuEbdS9+V5/mggEM32l4rXrTyze2dIUBkxb11Oap
/20kg0kQYgLm9FlfNNrH5TyAyQKBgQDG53C9FAiS5Jvp4mHvp3Mw2WLSxLOGVsI9
9PWzAbm5J/Iyh0o3aw+AxicSZJQVDuKnwb9nN9tY3rSRSuV7b1MZWNT1DVSRxJv4
fLT6x0Thj8G18dRrQk47fcpSBXJZgoA3p6uYJ7NAaor6/M+BaRKLFaLnI5VUBLCI
g0EJsFqZkQKBgHGGviz7tj/50LmGYSRMHWB+3bGXBl4YCN68c5VYfYcs9YdiB+Wh
XiXYVhKkowd7tdT65J4WAy+sQHpOvRliZ3yD6nG6AOm1bKcG0D5pxOdFz7doiOWU
Y0eUI8dIj6po3Q0fe5tdaEQuwMnnBLhv4FFvOHQEiXA/H+8C1vyDZ7J5AoGBAL7Z
ZWhXJ57iseNpr3UHY9S0UUlRKby11G0kdPhMqIrLlG39OC4SkO6LswPH8/p8rOgw
VFpS0ynjjrh2d9jHSaHhhXyYTNQNJWb5KaISv7n6ePD+WEaOeYhFu3KV/VF7kVFV
laVKOomNcUtAdk6CpiNEIGnNz8OtOEHURvxn9ZaxAoGBAMAsUTNGZxye7H50WOWE
7jZGbC1w4ZCvJTTR8V9ssUNtKdO1LAtvalxltdY34ITfJuGsfKBM3WTuJrgAxwLl
75UvQt8JuzI85CsntfTUlgc831fOUp9Vx3mvv0rQbf6RfqcM/rAYOvkZohz/zdN4
BjSm04g7YL7FphytAR3hqdQe
-----END PRIVATE KEY-----
```

input:
```
TCaMaZiybTdGVGUYlQu79lzdS3JACWFw586fJB8w5NQ0wS4aTYnq3Z15RpeZ9AwjrO/uiFWSO2+z
3yqfyF1cqkQ12PacaduR9uv2R77Sr4UdfpKvxQ//FedUKMV3re1E8ruqnSpmNqU2YUZZcU88UvE3
KLAz684IvnNAH6/JOw8/Mh+JhiBDu12L4qWtT8+9sKLDRmoSR9GEatVUDfotKTt0zKoMsPGr3b0W
0Z7HARSv0QTwEzveE3UUl8gbq+8gUxOWk3FJHzvsUJlLKg942TKJoh/yh1z38ARfW2LvawaYZUhB
h3RssGbDH0/YoX0V3HyjVyofqbg9hfIqvpx/hQ==
```

Reply:
I don't understand how you're getting specific ciphertext with 3.0.9. When I compiled 3.0.9 from sources I'm getting the expected padding check failed message:
```
~/workspace/openssl/apps/openssl rsautl -inkey key.pem -decrypt -in ciphertext.bin -pkcs -out plaintext.bin
The command rsautl was deprecated in version 3.0. Use 'pkeyutl' instead.
RSA operation error
40874182A57F0000:error:0200009F:rsa routines:RSA_padding_check_PKCS1_type_2:pkcs decoding error:crypto/rsa/rsa_pk1.c:269:
40874182A57F0000:error:02000072:rsa routines:rsa_ossl_private_decrypt:padding check failed:crypto/rsa/rsa_ossl.c:499:
```
```
~/workspace/openssl/apps/openssl version
OpenSSL 3.0.9 30 May 2023 (Library: OpenSSL 3.0.9 30 May 2023)
```

When I try the 3.2.0 release I indeed get the same message as you did.

If I had to guess, I'd say that the issue is mismatched keys: one of the reasons why I'm checking for 48 byte long result, is if the key doesn't match the ciphertext then the output won't be 48 byte long. And it is 48 byte long only because the generator has the access to the private key. So if the key was recreated after the `pms_values.bin` file was created, it will not decrypt to 48 bytes.

Reply:
Filed #8 to make debugging this easier, but almost certainly won't work on it till next year.

Reply:
> The implicit rejection support must have been backported to OpenSSL 3.0.9 on Fedora.

no, it's not backported to Fedora:
```
$ rpm -q openssl
openssl-3.0.9-1.fc37.x86_64
$ openssl rsautl -inkey key.pem -decrypt -in ciphertext.bin -pkcs -out plaintext.bin
The command rsautl was deprecated in version 3.0. Use 'pkeyutl' instead.
RSA operation error
007E558DC47F0000:error:0200009F:rsa routines:RSA_padding_check_PKCS1_type_2:pkcs decoding error:crypto/rsa/rsa_pk1.c:269:
007E558DC47F0000:error:02000072:rsa routines:rsa_ossl_private_decrypt:padding check failed:crypto/rsa/rsa_ossl.c:499:
```

But the implicit rejection does depend on the calculated value `d` (the private exponent) so if it has a different value then the returned message will be different. It probably decrypts correctly as if `p` and `q` are present OpenSSL will use the CRT optimisation, where it doesn't even look at `d`, but then it will use `d` for the actual implicit rejection: thus difference in the overall API call.

(I've purposefully used `d` as the secret for the implicit rejection KDF as that makes it simpler for the the differences between two-prime and multi-prime RSA keys)

Reply:
If the 48-bytes-return-on-decryption-failure is a feature due to implicit rejection, then at least for some reason it's there in 3.0.9:
```
[stefanb@... openssl-3.0.9-2.fc38.x86_64]$ pwd
usr/src/debug/openssl-3.0.9-2.fc38.x86_64
[stefanb@... openssl-3.0.9-2.fc38.x86_64]$ grep -r implicit_re ./
./crypto/cms/cms_env.c:        EVP_PKEY_CTX_ctrl_str(ktri->pctx, "rsa_pkcs1_implicit_rejection", "0");
./crypto/evp/ctrl_params_translate.c:      "rsa_pkcs1_implicit_rejection",
./crypto/pkcs7/pk7_doit.c:        EVP_PKEY_CTX_ctrl_str(pctx, "rsa_pkcs1_implicit_rejection", "0");
[...]
```

Anyway, I think this test script is now in decent shape.

Reply:
ack. I'm super busy right now, so won't be able to do a review it deserves righ away, but on quick look it looks fine, so feel free to use it. Will do proper review later (likely next year, sorry)

Reply:
To finish this discussion: yes implicit rejection on F38 is expected, it was backported there, it wasn't backported to the F37 I was using at the time. So it looks to me like expected behaviour